subprojects/packagefiles/highs/meson.build

```diff
@@ -0,0 +1,102 @@
+project('HiGHS', ['c', 'cpp'],
+  version : '1.2.0',
+  default_options : ['warning_level=3', 'cpp_std=c++11'])
+
+compiler = meson.get_compiler('cpp')
+
+have_mm_pause = compiler.compiles('''
+  #include <immintrin.h>
+  int main () {
+    _mm_pause();
+    return 0;
+  }''',
+  name : '_mm_pause')
+
+have_builtin_clz = compiler.compiles('''
+  #include <cstdint>
+  int main () {
+    unsigned int x = 5;
+    unsigned long long y = __builtin_clz(x);
+    x = __builtin_clzll(y);
+    return 0;
+  }''',
+  name : '__builtin_clzll')
+
+have_bitscan_reverse = compiler.compiles('''
+  #include <intrin.h>
+  #pragma intrinsic(_BitScanReverse)
+  #pragma intrinsic(_BitScanReverse64)
+  int main () {
+    unsigned long x = 5;
+    unsigned long y;
+    _BitScanReverse(&y, x);
+    _BitScanReverse64(&x, y);
+    return 0;
+  }''',
+  name : '_BitScanReverse')
+  
+highs_include_directories = include_directories([
+  '.',
+  'app',
+  'extern',
+  'src',
+  'src/io',
+  'src/ipm/ipx',
+  'src/ipm/ipx/include/',
+  'src/ipm/basiclu',
+  'src/lp_data',
+  'src/mip',
+  'src/model',
+  'src/presolve',
+  'src/qpsolver',
+  'src/simplex',
+  'src/test',
+  'src/util',
+])
+
+ipx_on = get_option('ipx')
+if not get_option('fast_build')
+  ipx_on = true
+endif
+
+version_split = meson.project_version().split('.')
+conf_data = configuration_data()
+conf_data.set('FAST_BUILD', get_option('fast_build'))
+#conf_data.set('SCIP_DEV', scip_dev)
+#conf_data.set('HiGHSDEV', highs_dev)
+conf_data.set('IPX_ON', ipx_on)
+conf_data.set('HIGHSINT64', get_option('highsint64'))
+conf_data.set('CMAKE_SOURCE_DIR', meson.project_source_root())
+conf_data.set('HIGHS_VERSION_MAJOR', version_split[0])
+conf_data.set('HIGHS_VERSION_MINOR', version_split[1])
+conf_data.set('HIGHS_VERSION_PATCH', version_split[2])
+conf_data.set('HIGHS_HAVE_MM_PAUSE', have_mm_pause)
+conf_data.set('HIGHS_HAVE_BUILTIN_CLZ', have_builtin_clz)
+conf_data.set('HIGHS_HAVE_BITSCAN_REVERSE', have_bitscan_reverse)
+
+hconfig = configure_file(
+  input : 'src/HConfig.h.in',
+  output : 'HConfig.h',
+  configuration : conf_data,
+  format : 'cmake@') 
+               
+hconfig_include_dir = include_directories('.')
+
+subdir('src')
+highs_deps = [dependency('threads')]
+if ipx_on
+  highs_deps += ipx_dep
+endif
+
+libhighs = library(
+  'highs', highs_sources,
+  dependencies : highs_deps,
+  include_directories : highs_include_directories) 
```

Comment:
I assume this is probably installable... ?

Reply:
What is the policy on installing things if they are a subproject? I remember there was once the problem that such libraries will be installed along with their consuming projects and then collide with system-wide installations of the library.

Is this no longer a problem or do I need to include some check?

Reply:
A check would look something like

```
not meson.is_subproject() or get_option('default_library') == 'static'
```

i.e. only install if highs is being built standalone as its own project, OR if it's configured to build and link as a shared library which must be available at runtime.

Meson did recently add support for a manual `meson install` option that only installs runtime files -- i.e. shared libraries but not static libraries. This CANNOT be handled per-subproject but if the main project doesn't install any devel files anyway it doesn't matter. It's an imperfect solution but may be good enough for you. In that case, just install and don't bother with checks.

Which one do you prefer? My usual approach is to defer the decision to the people who are affected, since neither solution seems wrong to me.

Reply:
This is controversial topic. I very strongly disagree with @eli-schwartz, IMHO there should be no special case for subproject, they should install everything just like any project. What user actually want is to not install devel files, such as static libraries, and that can be done using meson install --tags.

What @eli-schwartz suggest is completely inconsistent with every other meson projects.

Reply:
> Which one do you prefer? My usual approach is to defer the decision to the people who are affected, since neither solution seems wrong to me.

The decision is not for the maintainer to make, it is fundamentally a user decision what they install.

Reply:
I do not have a strong opinion on this, so I will skip the check.

Reply:
I have explained ad nauseam that users ***cannot*** decide, the existing options are insufficient and you're coming from the perspective of someone that primarily uses subprojects in the opposite way from what the wrapdb intends. So it's no surprise that you disagree with me.

I'm grateful that the situation isn't as bad as it used to be. But it's by no means perfect.

Your disagreement is duly noted. Meanwhile I will continue to prioritize the needs of actual users rather than validation of one's own proposed solution that is insufficient for user needs.

Reply:
@lukas-weber great, that makes things wonderfully simple.

Reply:
> Meanwhile I will continue to prioritize the needs of actual users rather than validation of one's own proposed solution that is insufficient for user needs.

I have absolutely have no idea at all what is missing for users, please provide suggestions in a separate issue.